File: top-k-frequent-elements/eunhwa99.java
import java.util.HashMap;
import java.util.Map;
import java.util.PriorityQueue;

// 시간 복잡도: O(nlogk) - 최대 힙에 k개의 요소를 넣는데 O(logk)가 걸리고, 이를 n번 반복
// 공간 복잡도: O(n) - 빈도수를 저장하는 freqMap
class Solution {

  public int[] topKFrequent(int[] nums, int k) {
    Map<Integer, Integer> freqMap = new HashMap<>();
    for (int num : nums) {
      freqMap.put(num, freqMap.getOrDefault(num, 0) + 1);
    }

    PriorityQueue<Map.Entry<Integer, Integer>> maxHeap =
        new PriorityQueue<>((a, b) -> b.getValue() - a.getValue());

    // 빈도수를 기준으로 최소 힙에 k개의 요소를 넣기
    for (Map.Entry<Integer, Integer> entry : freqMap.entrySet()) {
      maxHeap.offer(entry);
    }

    int[] result = new int[k];
    int index = 0;
    while (k > 0) {
      k--;
      result[index++] = maxHeap.poll().getKey();
    }

    return result;
  }
}

File: find-median-from-data-stream/eunhwa99.kt
import java.util.PriorityQueue
class MedianFinder() {
    // 1 2 3 4 5
    // 중간값을 기준으로 
    //   3
    //  2 4
    // 1   5
    // 1 2 : maxHeap
    // 4 5 : minHeap
    private val maxHeap = PriorityQueue<Int> { a, b -> b - a }
    private val minHeap = PriorityQueue<Int>()

    // TC: O(log n)
    // SC: O(n)
    fun addNum(num: Int) {
       if(maxHeap.isEmpty() || maxHeap.peek() >= num) {
        maxHeap.add(num)
       } else {
        minHeap.add(num)
       }

       if(maxHeap.size > minHeap.size + 1) {
        minHeap.add(maxHeap.poll())
       } else if(minHeap.size > maxHeap.size) {
        maxHeap.add(minHeap.poll())
       }
    }

    fun findMedian(): Double {
       if(maxHeap.size == minHeap.size) {
        return (maxHeap.peek() + minHeap.peek()) / 2.0
       } else {
        return maxHeap.peek().toDouble()
       }
    }

}
